src/uct/api/uct.h

```diff
@@ -881,13 +949,130 @@ struct uct_ep_params {
 
     /**
      * Callback that will be used for filling the user's private data to be
-     * delivered to the server by @ref uct_sockaddr_conn_request_callback_t.
+     * delivered to the remote peer by the callback on the server or client side.
      * This field is only valid if @ref uct_ep_params_t::sockaddr is set.
      * @note It is never guaranteed that the callaback will be called. If, for
      * example, the endpoint goes into error state before issuing the connection
      * request, the callback will not be invoked.
      */
     uct_sockaddr_priv_pack_callback_t sockaddr_pack_cb;
+
+    /**
+     * The connection manager object as created by @ref uct_cm_open.
+     * Either @a cm or @a iface field must be initialized but not both.
+     */
+    uct_cm_h                          cm;
+
+    /**
+     * Connection request what was passed to
+     * @ref uct_listener_conn_request_callback_t .
+     */
+    uct_conn_request_h                conn_request;
+
+    union {
+        /**
+         * Callback that will be invoked when the endpoint on the client side
+         * is being connected to the server by a connection manager @ref uct_cm_h .
+         */
+        uct_ep_client_connect_cb_t      client;
+
+        /**
+         * Callback that will be invoked when the endpoint on the server side
+         * is being connected to a client by a connection manager @ref uct_cm_h .
+         */
+        uct_ep_server_connect_cb_t      server;
+    } sockaddr_connect_cb;
+
+    /**
+     * Callback that will be invoked when the endpoint is disconnected.
+     */
+    uct_ep_disconnect_cb_t              disconnect_cb;
+};
+
+
+/**
+ * @ingroup UCT_RESOURCE
+ * @brief Connection manager attributes, capabilities and limitations.
+ */
+struct uct_cm_attr {
+    /**
+     * Mask of valid fields in this structure, using bits from
+     * @ref uct_cm_attr_field. Fields not specified by this mask
+     * will be ignored.
+     */
+    uint64_t    field_mask;
+
+    /**
+     * Max size of the connection manager's private data used for connection
+     * establishment with sockaddr.
+     */
+    size_t      max_conn_priv;
+};
+
+
+/**
+ * @ingroup UCT_RESOURCE
+ * @brief Parameters for creating a connection manager object @ref uct_cm_h by
+ * @ref uct_cm_open
+ */
+struct uct_cm_params {
+    /**
+     * Mask of valid fields in this structure, using bits from
+     * @ref uct_cm_params_field. Fields not specified by this mask
+     * will be ignored.
+     */
+    uint64_t                          field_mask;
+
+    /**
+     * UCT component, mandatory field.
+     */
+    uct_component_h                   component;
```

Comment:
@yosefe why we don't pass mandatory field as a separate argument ? (this is true for other parts of code as well)

Reply:
@shamisp for example `iface` was a mandatory field for `uct_ep_create`, but now we are adding `cm` object which is an alternative for `iface`.

Reply:
today this field is mandatory, but in order be be future compatible, and allow making this field non-mandatory sometime, it is part of the struct

Reply:
@yosefe there is value in passing this as an argument. If we want to make a generic interface we just can pass `void *` everywhere. I think mandatory and conceptual object like iface and ep have to be passed as a separate argument.